src/Psalm/Type/Reconciler.php

```diff
@@ -124,7 +124,7 @@ public static function reconcileKeyedTypes(
 
             $cloned_referenceds = [];
             foreach ($existing_references as $reference => $referenced) {
-                if (!isset($cloned_referenceds[$referenced])) {
+                if (!isset($cloned_referenceds[$referenced]) && isset($old_existing_types[$referenced])) {
```

Comment:
@AndrolGenhald I'm not sure whether this is enough, or
* whether there should be a fallback union type assigned to `$existing_types[$referenced]`, or
* whether the `$referenced` value (e.g. `'$doesNotMatter[$a]'`) should to be split and processed

Reply:
I'll step through it with your test tonight or tomorrow and see what's going on. Thanks for the reproduction and possible fix!

Reply:
Seems like it needs to have a type, I just ran into another error in a project having this patch applied:
```
Uncaught Error: Call to a member function equals() on null in vendor/vimeo/psalm/src/Psalm/Context.php:524
```
https://github.com/vimeo/psalm/blob/a32e63f131466a8965a47b58649bc72daeaf8b25/src/Psalm/Context.php#L521

Reply:
Oh, yeah, having nulls in there is not a good idea. I think I know what's going on though; when we do the check against `$a` it potentially changes `$a`'s type, which means we remove everything dependent on `$a` from `Context::$vars_in_scope` (somewhere, can't remember where right now) because `$doesNotMatter[$a]` doesn't have the same type if `$a` has changed. When that happens, we also need to do _something_ to any references that exist to `$doesNotMatter[$a]`, probably by changing whatever code removes `$doesNotMatter[$a]` to use the normal process by calling `Context::remove()` instead of whatever it's doing now.

Of course this means the fact that it's a reference will be lost, but like I said before, combining references with arrays and objects is just asking for trouble, and there's not much we can do about it.